source/agora/consensus/EnrollmentManager.d

```diff
@@ -568,6 +569,120 @@ public class EnrollmentManager
         return true;
     }
 
+    /***************************************************************************
+
+        Check if a pre-image exists
+
+        Params:
+            enroll_key = The key for the enrollment in which the pre-image is
+                contained.
+            height = The block height of the preimage to check existence
+
+        Returns:
+            true if the pre-image exists
+
+    ***************************************************************************/
+
+    public bool hasPreimage (const ref Hash enroll_key, ulong height) @trusted
+    {
+        auto results = this.db.execute("SELECT preimage from validator_set" ~
+                    " WHERE key = ?", enroll_key.toString());
```

Comment:
The `toString()` call is allocating. I think something like this could work:

```D
static char[] cmd;
cmd.length = 0;
assumeSafeAppend(cmd);
sformat(cmd, "SELECT preimage from validator_set WHERE key = %s", enroll_key);
auto results = this.db.execute(cmd);
```